src/rp2/abstract_accounting_method.py

```diff
@@ -41,11 +47,17 @@ def __init__(
     ) -> None:
         self.__accounting_method: AbstractAccountingMethod = accounting_method
         self.__acquired_lot_list = acquired_lot_list
+        self.__acquired_lot_heap: List[Tuple[Union[float, RP2Decimal], InTransaction]] = []
         self.__acquired_lot_2_partial_amount = acquired_lot_2_partial_amount
         self.__to_index = 0
         self.__from_index = 0
 
     def set_to_index(self, to_index: int) -> None:
+        # Control how far to advance the iterator, caller is responsible for updating
+        if self.__accounting_method.use_heap():
```

Comment:
I'm thinking rather than having a conditional on `use_heap`, can we make this method abstract and implement it in the subclasses? If I'm correct that should allow us to:
* remove `use_heap` altogether;
* move `add_selected_lot_to_heap` out of the abstract class and into hifo and lifo as a private method (or maybe into a `AbstractHeapAccountingMethod` superclass);
* move  `heap_key` out of the abstract class and into hifo and lifo as a private method (or maybe into a `AbstractHeapAccountingMethod` superclass);
* perhaps move `ListAccountingMethodIterator` and `HeapAccountingMethodIterator` in the subclasses too.

This would have the advantage of keeping this superclass free of implementation details.

Reply:
I've got all of your requests implemented other than:
> I'm thinking rather than having a conditional on use_heap, can we make this method abstract and implement it in the subclasses?

Just to clarify, this is currently on `LotCandidates` so no subclasses (unless you mean move all of this logic into `AbstractHeapAccountingMethod`)

> - move add_selected_lot_to_heap out of the abstract class and into hifo and lifo as a private method (or maybe into a AbstractHeapAccountingMethod superclass);

This is somewhat coupled to the above and not seeing an obvious solution. I'll think about it a bit more Sunday.

Reply:
Good point. One possible solution:
* `LotCandidates` becomes a hierarchy of classes (parallel to `AbstractAccountingMethod`)
* `AbstractAccountingMethod` gets a new `create_lot_candidates()` abstract method, which gets implemented in the subclasses
* the [instantiation of LotCandidates](https://github.com/eprbell/rp2/blob/main/src/rp2/accounting_engine.py#L122) in `AccountingEngine` becomes `node.value.create_lot_candidates()`

However before we go too far down the refactoring rabbit hole, let's get this PR to work: we can reevaluate how much we want to refactor at the end.